### PR TITLE
Added Hough Line Transform

### PR DESCRIPTION
**Pull request by annimesh2809**
Added a new feature detection : Hough line transform as mentioned in #600 .

The method returns an array of tuples containing (r,theta) as parametric values of lines present in the image.

### CODE REVIEW

File: src/houghtranform.jl [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            rho::Float64, theta::Float64,
+            min_theta::Float64, max_theta::Float64,
+            threshold::Int64, linesMax::Int64)
+   
+    theta > 0 || error("theta threshold must be positive")
+    rho > 0 || error("radius threshold must be positive")
+    min_theta < max_theta || error("max_theta must be greater than min_theta")
+
+    if ! is_binary_image(img)
+        img = canny(img)
```
Comment (timholy) [inline src/houghtranform.jl]: This breaks orthogonality; what if someone wants a different algorithm than `canny`? Will be solved if we simply require a `Bool` eltype.

Hmm, I hadn't noticed that `canny` returns an `Array{Float64}`. We should probably change it to return an `Array{Bool}`.

File: src/houghtranform.jl [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            img::AbstractArray{T,2},
+            rho::Float64, theta::Float64,
+            min_theta::Float64, max_theta::Float64,
+            threshold::Int64, linesMax::Int64)
+   
+    theta > 0 || error("theta threshold must be positive")
+    rho > 0 || error("radius threshold must be positive")
+    min_theta < max_theta || error("max_theta must be greater than min_theta")
+
+    if ! is_binary_image(img)
```
Comment (timholy) [inline src/houghtranform.jl]: Better to just require an `T<:Union{Bool,Gray{Bool}}` eltype.

File: src/houghtranform.jl [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        img = canny(img)
+    end
+
+    width::Int64 = size(img)[2]
```
Comment (timholy) [inline src/houghtranform.jl]: No need to declare types like this.

File: src/houghtranform.jl [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    min_theta   = minimum theta of lines
+    max_theta   = maximum theta of lines
+    threshold   = no of points to pass through line for considering it valid
+    linesMax    = maximum no of lines to return
+
+"""
+
+function hough_transform_standard{T}(
+            img::AbstractArray{T,2},
+            rho::Float64, theta::Float64,
```
Comment (timholy) [inline src/houghtranform.jl]: Too restrictive on the types? Particularly for people who have 32-bit machines, `Int64` is not native (better to use `Int`). `Number` and `Integer` might even be best.

File: src/houghtranform.jl [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    min_theta < max_theta || error("max_theta must be greater than min_theta")
+
+    if ! is_binary_image(img)
+        img = canny(img)
+    end
+
+    width::Int64 = size(img)[2]
+    height::Int64 = size(img)[1]
+    irho::Float64 = 1 / rho;
+    numangle::Int64 = round((max_theta - min_theta)/theta)
```
Comment (timholy) [inline src/houghtranform.jl]: `round(Int, x)` returns an `Int`. You can use this in several places below, too.

File: src/houghtranform.jl [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    numrho::Int64 = round(((width + height)*2 + 1)/rho)
+
+    accumulator_matrix = Array{Int64}(numangle + 2, numrho + 2)
+    fill!(accumulator_matrix, 0)
+
+    #Pre-Computed sines and cosines in tables
+    tabsin = Array{Float64}(numangle)
```
Comment (timholy) [inline src/houghtranform.jl]: Even better, `Vector{Float64}` (here and below).

File: src/houghtranform.jl [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    tabcos = Array{Float64}(numangle)
+    for i in 1:numangle
+        tabsin[i] = (sin(min_theta + (i-1)*theta) * irho)
+        tabcos[i] = (cos(min_theta + (i-1)*theta) * irho)
+    end
+
+
+    #Hough Transform implementation
+    for pix in CartesianRange(size(img))
+        if img[pix] == 1.0
```
Comment (timholy) [inline src/houghtranform.jl]: If you're using `Bool` eltypes, this becomes just `if img[pix]`

File: src/houghtranform.jl [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            for i in 1:numangle
+                dist::Int64 = round((pix[1] * tabsin[i] + pix[2] * tabcos[i]))
+                dist += round((numrho -1)/2)
```
Comment (timholy) [inline src/houghtranform.jl]: This constant could be hoisted out of the loop.

File: src/houghtranform.jl [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                accumulator_matrix[i + 1, dist + 1] += 1
+            end
+        end
+    end
+
+    #Finding local maximum lines
+    validLines = Array{CartesianIndex}(0)
+    for val in CartesianRange(size(accumulator_matrix))
```
Comment (timholy) [inline src/houghtranform.jl]: Perhaps we should add a function input to `findlocalmaxima` and just use that?

File: src/houghtranform.jl [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            accumulator_matrix[val] > accumulator_matrix[val[1] - 1,val[2]] &&
+            accumulator_matrix[val] >= accumulator_matrix[val[1] + 1,val[2]]
+            push!(validLines,val)
+        end
+    end
+
+    #Sorting by value in accumulator_matrix
+    sort!(validLines, by = (x)->accumulator_matrix[x], rev = true)
+
+    linesMax = min(linesMax, size(validLines)[1])
```
Comment (timholy) [inline src/houghtranform.jl]: can just use `length` instead of `size(validLines)[1]` (they're equivalent if `validLines` is a vector).

**Comment by annimesh2809** [comment]
Is there any function available for drawing lines using the normal parameters generated by hough transform? I found the package Cairo.jl but is there something simpler than that.

There is an ImageDraw.jl package and a line function implemented in it. Should I write a function to draw line from normal parameters which uses this existing function?

**Comment by annimesh2809** [comment]
@timholy Yes, I also think it would be better to have it in ImageFeatures.jl instead of having it directly in Images. Should I create a new pull request in that repo?  Also what updates are required in the ImageFeatures.jl package related to the new images framework, maybe I can help :)

**Comment by timholy** [comment]
@annimesh2809, I suspect ImageDraw would indeed be the better way to go...fewer dependencies.

Great, I think ImageFeatures is the best place. With regards to updating ImageFeatures, if you install it and type `using ImageFeatures` you'll already notice a couple of deprecation warnings that should be fixed. The next step would be to run the tests and see if there are more depwarns or breakages. Probably just getting it updated should be a PR on its own, then new stuff can be added.

**Comment by mronian** [comment]
@annimesh2809 It would be great if you work on updating ImageFeatures to the new Images.jl. I have been wanting to do that for a long time but haven't been able to get time. I'd be happy to help in case you need any 😄

**Comment by annimesh2809** [comment]
@mronian I would be really happy to update ImageFeatures.jl. But I did not understand what you meant by update ImgaeFeatures to the new Images.jl

**Comment by timholy** [comment]
The deprecation warnings and test errors should basically tell you what to do.

**Comment by timholy** [comment]
I realized that it may not be obvious that to you "the new Images" *is* what you're running now. There was a big revamp described in #542 and merged in #577. ImageFeatures was written before that work was even started.

**Comment by mronian** [comment]
These links might help -
- https://github.com/JuliaImages/Images.jl/blob/master/NEWS.md
- http://juliaimages.github.io/latest/
- #577

**Comment by annimesh2809** [comment]
I have started working on updating ImageFeatures.jl to work with the new images. But there are some failing test cases without changing any code:

img1
img2

Should I comment out those test cases and create a log of them so that we can correct those test cases later when all the deprecated warnings have been removed?
Also should I create a new issue on ImageFeatures.jl so we can discuss the progress of this update?

**Comment by mronian** [comment]
I guess some changes to Images.jl have modified some values of the descriptor. This should be solved if take an approximate equals instead of equals. Lets close this thread and move it to ImageFeatures.jl.

**Comment by mronian** [comment]
@annimesh2809 Lets continue this discussion at JuliaImages/ImageFeatures.jl#24

**Comment by mronian** [comment]
Whoops. I did not realise this was a PR for hough lines. I thought it was just an issue. I'll reopen it.

**Comment by annimesh2809** [comment]
@mronian Actually we should close this PR because hough lines will be added to ImageFeatures.jl as discussed above once we remove all the errors from that package.

**Pull request closed without merging**